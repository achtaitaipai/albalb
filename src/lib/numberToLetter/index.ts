import { words } from "./words";

const suffix = {
  mille: 2,
  million: 3,
};

const cleanNumber = (str: string | number) =>
  typeof str === "string"
    ? parseInt(str).toString()
    : Math.round(str).toString();

const splitNumber = (str: string) => {
  const reversed = str.split("").reverse().join("");
  const chuncks = reversed.match(/(\d{3}|\d+)/g) ?? [];
  const res = chuncks.map((c) => c.split("").reverse()).reverse();
  return res;
};

export const numberToLetter = (input: number | string) => {
  const str = splitNumber(cleanNumber(input));
  const chunks = str.map((chunk, i) => {
    const num = parseInt(chunk.join(""));
    if (num === 0) return "";
    if (num < 100) return words[num];
    const c = Math.floor(num / 100);
    const d = parseInt(chunk.slice(1).join(""));
    const start = c > 1 ? words[c] + "-cent" : "cent";
    const end = d > 0 ? "-" + words[d] : "";
    const res = start + end;
    if (end === "" && i === str.length - 1 && c > 1) return res + "s";
    return res;
  });
  Object.entries(suffix).forEach(([toAdd, i]) => {
    for (let index = chunks.length - i; index >= 0; index -= i) {
      if (chunks[index] === "") return;
      chunks[index] =
        i === 0 || chunks[index] !== "un" ? chunks[index] + "-" + toAdd : toAdd;
    }
  });
  return chunks.filter((c) => c !== "").join("-");
};
