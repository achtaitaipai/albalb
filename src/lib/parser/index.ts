import { solve } from ".";
import { compile } from ".";
import { tokenize } from ".";

export { tokenize } from "./tokenize";
export { compile } from "./compile";
export { solve } from "./solve";

export const parse = (input: string) => solve(compile(tokenize(input)));
