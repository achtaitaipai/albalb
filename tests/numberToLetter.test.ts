import { describe, it, expect } from "vitest";
import { numberToLetter } from "../src/lib/numberToLetter";

describe("numberToLetter()", () => {
  it("returns the correct number", () => {
    expect(numberToLetter(1)).toMatchSnapshot();
    expect(numberToLetter(21)).toMatchSnapshot();
    expect(numberToLetter(100)).toMatchSnapshot();
    expect(numberToLetter(200)).toMatchSnapshot();
    expect(numberToLetter(243)).toMatchSnapshot();
    expect(numberToLetter(1000)).toMatchSnapshot();
    expect(numberToLetter(11000)).toMatchSnapshot();
    expect(numberToLetter(863969)).toMatchSnapshot();
    expect(numberToLetter(146548978)).toMatchSnapshot();
  });
});
