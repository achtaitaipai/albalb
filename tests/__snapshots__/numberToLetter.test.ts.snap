// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`numberToLetter() > returns the correct number 1`] = `"un"`;

exports[`numberToLetter() > returns the correct number 2`] = `"vingt-et-un"`;

exports[`numberToLetter() > returns the correct number 3`] = `"cent"`;

exports[`numberToLetter() > returns the correct number 4`] = `"deux-cents"`;

exports[`numberToLetter() > returns the correct number 5`] = `"deux-cent-quarante-trois"`;

exports[`numberToLetter() > returns the correct number 6`] = `"mille"`;

exports[`numberToLetter() > returns the correct number 7`] = `"onze-mille"`;

exports[`numberToLetter() > returns the correct number 8`] = `"huit-cent-soixante-trois-mille-neuf-cent-soixante-neuf"`;

exports[`numberToLetter() > returns the correct number 9`] = `"cent-quarante-six-million-cinq-cent-quarante-huit-mille-neuf-cent-soixante-dix-huit"`;
